### PR TITLE
Add weakref processing work packet

### PR DESCRIPTION
**Pull request by wenyuzhao**
This PR:

1. Adds `ProcessWeakRefs` work packet that will call into vm-binding to do complex weak reference processing jobs.
2. Adds code to support ephemeron objects processing.

### CODE REVIEW

**Review by qinsoon — COMMENTED** [review]
This PR adds a VM-dependent approach for weak reference processing on top of the existing Java-style reference processing (which is unused at the moment). I assume a binding would only use one of the two approaches, right? Should we provide a feature or an option for the binding to choose which approach to use?

File: src/plan/gencopy/global.rs [hunk 1]
```diff
@@ -123,6 +123,8 @@ impl<VM: VMBinding> Plan for GenCopy<VM> {
         // Prepare global/collectors/mutators
         scheduler.work_buckets[WorkBucketStage::Prepare]
             .add(Prepare::<Self, GenCopyCopyContext<VM>>::new(self));
+        scheduler.work_buckets[WorkBucketStage::RefClosure]
+            .add(ProcessWeakRefs::<GenCopyMatureProcessEdges<VM>>::new());
```
[the remark this answers is not in the document]
Comment (wenyuzhao) [reply src/plan/gencopy/global.rs]: Fixed. V8 cannot do real nursery gc right now. But yes I guess it depend on nursery/mature GC. I've fixed this and use GenCopyNurseryProcessEdges for nursery GCs,

File: src/scheduler/scheduler.rs [hunk 2]
```diff
@@ -144,6 +166,10 @@ impl<C: Context> Scheduler<C> {
         *self.finalizer.lock().unwrap() = w.map(|w| box w as Box<dyn CoordinatorWork<C>>);
     }
 
+    pub fn on_closure_end(&self, f: Box<dyn Send + Fn() -> bool>) {
```
[the remark this answers is not in the document]
Comment (wenyuzhao) [reply src/scheduler/scheduler.rs:169]: Fixed

**Comment by wenyuzhao** [comment]
> This PR adds a VM-dependent approach for weak reference processing on top of the existing Java-style reference processing (which is unused at the moment). I assume a binding would only use one of the two approaches, right? Should we provide a feature or an option for the binding to choose which approach to use?

Probably this is not necessary? If the VM does not use one of the approaches, then it will become a no-op and do nothing. Actually its harmless for a VM to choose both approaches, if it really need to do so.

**Review by qinsoon — APPROVED** [review]
LGTM

**Pull request merged**